lessons/cryptography/cryptosystem-types/cryptosystems-types.md

```diff
@@ -0,0 +1,90 @@
+
+## Asymmetric Cryptosystems
+
+In an asymmetric cryptosystem, also known as a public-key cryptosystem, both the sender and receiver each have a private and a public key. Each person can generate their own private and public key pair. Each person should always keep their private key private, but can freely distribute their public key without any negative consequences. In the example below, Alice sends an encrypted message to Bob using an asymmetric cryptosystem. 
+
+![Sending a Message with Asymmetric Cryptography](.rsrc/asymmetric-crypto.png)
+
+In this example, Bob has shared his public key with Alice and keeps his private key private. Alice uses Bob's public key to encrypt a plaintext message, resulting in an encrypted ciphertext message. This encrypted message is sent to Bob. Bob then uses his private key to decrypt the ciphertext into Alice's original plaintext message. In this example, Alice's public and private keys are not needed since she is sending a message to Bob. If Bob wanted to send an encrypted message back to Alice, he would then need her public key, and Alice would use her private key to decrypt a message from Bob. 
+
+### RSA
+Rivest-Shamir-Adleman (RSA) is a type of asymmetric encryption that uses a private and public key based on two large prime numbers. The algorithm for encryption involves multiplying two very large prime numbers, and is considered very hard to crack because recovering the two original prime numbers is a mathematical problem that is currently thought to be infeasible to solve efficiently. There are some known attacks against RSA, but many of these attacks occur when small prime numbers are used. For example, if small prime numbers are used, an attacker could create a dictionary by encrypting high probability plaintext messages using various keys, and compare the resulting ciphertext with a stolen ciphertext message. If the messages match, the attacker now knows the original prime numbers used to encrypt the message. 
+
+### Diffie-Hellman
+The Diffie-Hellman key exchange process involves the use of asymmetric encryption, and is used to establish a shared private key for use in future communications. When we previously talked about AES (symmetric cryptosystem), we noted that Alice and Bob would need a way to securely share a private key. The Diffie-Hellman key exchange process can accomplish this. The Diffie-Hellman algorithm involves a public modulus and base value, and two secret values chosen by Alice and Bob. Through an elegant mathematical algorithm, Alice and Bob can each use their own private secret values to calculate a public key that can be sent to each other unencrypted. They can then each use the other person's public key and their own secret value to calculate the same secret key. 
```

Comment:
...calculate a public key that can be shared with anyone. Anyone can then use Alice's public key to encrypt a message that only she will be able to decrypt, with her private key. As long as the private key stays private to her, no one else will be able to decrypt that message.

Reply:
Done